### ISSUE TITLE
Can't fetch content behind paywall (monde-diplomatique.fr)

### ISSUE THREAD

**Issue by laBecasse** [wallabag/wallabag#6907]
Hello, 
I have the same issue as in  #3502, ie I can't the content of articles using the paywall for Monde Diplo. 

I checked the log of my server the authentication parameters are set by graby, but only the public content is retrieved. 

 - URL: https://www.monde-diplomatique.fr/2023/08/DE_CEUSTER/65977
 - wallabag version:  2.6.5

**Comment by HolgerAusB** [comment]
@laBecasse:  the login page named in vendor/j0ker/graby-site-config/monde-diplomatique.fr.txt is not available anymore. Try to change this file as follows. Maybe the last five login_extra_fields are not necessary anymore. Try it and report, so we can change this file for all. 
```
title: //h1[@class="h1"]
body: //div[contains(@class, 'contenu-principal')]
author: //meta[@property='article:author']/@content

strip: //div[contains(@class, 'bandeautitre')]
strip: //div[contains(@class, 'dates_auteurs')]

requires_login: yes

login_uri: https://www.monde-diplomatique.fr/connexion
login_username_field: email
login_password_field: mot_de_passe
login_extra_fields: valider=valider

## maybe remove the following extra_fields
login_extra_fields: page=connexion
login_extra_fields: formulaire_action=identification_sso
login_extra_fields: formulaire_action_args=@=xpath("//form//input[@name='formulaire_action_args']", request_html(config.getLoginUri()))
login_extra_fields: retour=http://www.monde-diplomatique.fr/
login_extra_fields: site_distant=http://www.monde-diplomatique.fr/


not_logged_in_xpath: //div[@id="paywall"]

test_url: https://www.monde-diplomatique.fr/1996/04/BOURDIEU/5425
```

**Comment by laBecasse** [comment]
Thanks for your answer. I tried both with and without the extra fields that could be removed, but it is not working.

**Comment by HolgerAusB** [comment]
Sorry, I once again forgot to mention, that you need to clear cache after each change of the config. Or did you do?

For Debian that's:
`sudo -u www-data /path-to-wallabag/bin/console --env=prod cache:clear`

**Comment by laBecasse** [comment]
Yes, I did. Sorry to not mention it before. 

I looked at the connexion form on the website :

```html
<form action="https://lecteurs.mondediplo.net?page=connexion_sso" method="post" id="identification_sso" >
	<div>
		<input name="page" value="connexion_sso" type="hidden">
		<input name="formulaire_action" type="hidden" value="identification_sso">
		<input name="formulaire_action_args" type="hidden" value="ECZkXUr9XO+2orHSJmTwMYjXPagiMy2NYwmfnaR8BwhsWQyH+j3cpmB3nl0BdQpKVstbDH0T6sx4JHHMqwd82c3Izzl67bo=">
		<input name="retour" value="https://www.monde-diplomatique.fr/mon_compte" type="hidden">
		<input name="site_distant" value="https://www.monde-diplomatique.fr/" type="hidden">
	</div>
	<ul>
		<li>
			<label>E-mail&nbsp;:</label>
			<input type="email" name="email" value="" autocapitalize="none" >
		</li>
		<li>
			<label>Mot de passe&nbsp;:</label>
			<input type="password" class="password " name="mot_de_passe" value="" >
			
		</li>
		<li class="editer_session">
			<div class="choix">
				<input type="checkbox" class="checkbox" name="session_remember" id="session_remember" value="oui" onchange="jQuery(this).addClass('modifie');">
				<label class="nofx" for="session_remember">Rester identifié quelques jours</label>
			</div>
		</li>
		<li class="submit">
			<input type="submit" name="valider" title="S'identifier" value="Valider">
		</li>
		<li class="dernier">
			<a title="Demander un nouveau mot de passe par courriel." href="https://boutique.monde-diplomatique.fr/customer/account/forgotpassword/" class="oubli">Mot de passe oublié ?</a>
		</li>
	</ul>
</form>
```

It seems like the  page, we use is not correct, but fixing it does not solve the issue.

**Comment by HolgerAusB** [comment]
@laBecasse I am no so good at paywall-sites and I can only guess. Your excerpt seem to be helpful. At first I thought the _uri should point to the site where the form is, but it could be the address where the form data is send to, so the original uri would be OK. But some of the form's field names and values differ slightly, so try:

```
login_uri: https://lecteurs.mondediplo.net?page=connexion_sso 
login_username_field: email
login_password_field: mot_de_passe

login_extra_fields: page=connexion_sso
login_extra_fields: formulaire_action=identification_sso
login_extra_fields: formulaire_action_args=@=xpath("//form//input[@name='formulaire_action_args']", request_html(config.getLoginUri()))
login_extra_fields: retour=http://www.monde-diplomatique.fr/mon_compte
login_extra_fields: site_distant=http://www.monde-diplomatique.fr/
login_extra_fields: valider=Valider

not_logged_in_xpath: //div[@id="paywall"]
```

maybe it is also a good idea to set a user_agent here, just in case mondediplo.net doesn't like the standadrd one:
`http_header(User-agent): Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/116.0`

**Comment by laBecasse** [comment]
I tried again your proposition with and without the changed user agent, but it is not working.

**Comment by HolgerAusB** [comment]
@laBecasse: please test with this one:

```
http_header(User-agent): Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/116.0
title: //h1[@class="h1"]
body: //div[contains(@class, 'contenu-principal')]
author: //meta[@property='article:author']/@content

strip: //div[contains(@class, 'bandeautitre')]
strip: //div[contains(@class, 'dates_auteurs')]

requires_login: yes

login_uri: https://www.monde-diplomatique.fr/sso/connexion/
login_username_field: email
login_password_field: mot_de_passe
login_extra_fields: valider=Valider

login_extra_fields: formulaire_action_args=@=xpath("//form//input[@name='formulaire_action_args']", request_html(config.getLoginUri()))

# also try:
#login_extra_fields: session_remember=oui

## don't see these in the form, might be old fields:
#login_extra_fields: page=connexion
#login_extra_fields: formulaire_action=identification_sso
#login_extra_fields: retour=http://www.monde-diplomatique.fr/
#login_extra_fields: site_distant=http://www.monde-diplomatique.fr/

not_logged_in_xpath: //div[@id="paywall"]

test_url: https://www.monde-diplomatique.fr/1996/04/BOURDIEU/5425
```

and also try to activate the session_remember line and/or the use of
`login_username_field: email_nobot` instead of `login_username_field: email`

**Comment by yomli** [comment]
Le Monde Diplomatique now requires a token (in an hidden input named `_jeton`, those extra fields you didn't see are in a div up the tree).

I managed to get a working version of `vendor/j0ker/graby-site-config/monde-diplomatique.fr.txt` but it's way slower than usual. I don't know how to optimize it:

```
http_header(User-agent): Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/116.0
title: //h1[@class="h1"]
body: //div[contains(@class, 'contenu-principal')]
author: //meta[@property='article:author']/@content

strip: //div[contains(@class, 'bandeautitre')]
strip: //div[contains(@class, 'dates_auteurs')]

requires_login: yes

login_uri: https://www.monde-diplomatique.fr/sso/connexion/
login_username_field: email
login_password_field: mot_de_passe
login_extra_fields: valider=Valider
login_extra_fields: session_remember=oui
login_extra_fields: formulaire_action=identification_lecteur_ws
login_extra_fields: retour=https://www.monde-diplomatique.fr/

login_extra_fields: formulaire_action_args=@=xpath("//form//input[@name='formulaire_action_args']", request_html(config.getLoginUri()))
login_extra_fields: _jeton=@=xpath("//form//input[@name='_jeton']", request_html(config.getLoginUri()))

## don't see these in the form, might be old fields:
#login_extra_fields: page=connexion
#login_extra_fields: email_nobot=
#login_extra_fields: site_distant=http://www.monde-diplomatique.fr/

not_logged_in_xpath: //div[@id="paywall"]

test_url: https://www.monde-diplomatique.fr/1996/04/BOURDIEU/5425
```

**Comment by fradav** [comment]
> Le Monde Diplomatique now requires a token (in an hidden input named `_jeton`, those extra fields you didn't see are in a div up the tree).
> 
> I managed to get a working version of `vendor/j0ker/graby-site-config/monde-diplomatique.fr.txt` but it's way slower than usual. I don't know how to optimize it:
> 
> ```
> http_header(User-agent): Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/116.0
> title: //h1[@class="h1"]
> body: //div[contains(@class, 'contenu-principal')]
> author: //meta[@property='article:author']/@content
> 
> strip: //div[contains(@class, 'bandeautitre')]
> strip: //div[contains(@class, 'dates_auteurs')]
> 
> requires_login: yes
> 
> login_uri: https://www.monde-diplomatique.fr/sso/connexion/
> login_username_field: email
> login_password_field: mot_de_passe
> login_extra_fields: valider=Valider
> login_extra_fields: session_remember=oui
> login_extra_fields: formulaire_action=identification_lecteur_ws
> login_extra_fields: retour=https://www.monde-diplomatique.fr/
> 
> login_extra_fields: formulaire_action_args=@=xpath("//form//input[@name='formulaire_action_args']", request_html(config.getLoginUri()))
> login_extra_fields: _jeton=@=xpath("//form//input[@name='_jeton']", request_html(config.getLoginUri()))
> 
> ## don't see these in the form, might be old fields:
> #login_extra_fields: page=connexion
> #login_extra_fields: email_nobot=
> #login_extra_fields: site_distant=http://www.monde-diplomatique.fr/
> 
> not_logged_in_xpath: //div[@id="paywall"]
> 
> test_url: https://www.monde-diplomatique.fr/1996/04/BOURDIEU/5425
> ```

Thanks it works now. I didn’t notice the slowing down you mentioned.

**Comment by HolgerAusB** [comment]
@laBecasse can you confirm, that this is working for you as well?

@yomli:
Thank you for figuring this out. ❤️

What do you mean with 'way slower'? We are talking about wallabag, which is just storing a page. It does not matter, how long it takes in my opinion, as long as there is no timeout while fetching. wallabag server needs to load the locked article first, only to find out that it is locked, then it must load the login page (or not?), then sending the credentials, wait for response and load the unlocked article and parse it afterward and reload own UI in your browser. I think, that could last around 3 to 6 seconds depending on the speed of your wallabag server, network and server load of monde-server. Of course it must be slower then just loading the article in your browser while already logged in. So if you think, that there is still a problem, which should be solved, you need to get more concrete, maybe in a new issue.

Do you want to contribute your config in [fivefilters/ftr-site-config](https://github.com/fivefilters/ftr-site-config)? If not, I can do that, if you want. This is the upstream repo for site-configs, used by graby+wallabag.

**Comment by yomli** [comment]
@HolgerAusB 
> What do you mean with 'way slower'? […] So if you think, that there is still a problem […]

No, I don't think that's a problem. I probably overreacted, mistakenly thinking that two `@=xpath` meant two processing of the same file (or worse, two requests, since `_jeton` don't change that often), which would slow everything down. But not knowing enough about how graby works, I'm probably wrong.

Since fradav didn't notice anything, I will submit the config in the upstream repo for site-configs.

> Additionally, please check, if that works for the English version, too.

Unfortunately, my account on the French version doesn't allow me to connect to the English version. Strangely enough, there doesn't seem to be a token (yet), so anyone able to test #7029 would hopefully have a better time figuring it out.

**Comment by HolgerAusB** [comment]
Thank you @yomli, I had hoped that your subscription works in English Edition too. Commit is merged.

@laBecasse please close this issue, if the patch works for you.

**Comment by TontonMollo** [comment]
Thanks, it works now on https://app.wallabag.it host.

**Comment by Gullumluvl** [comment]
Hi,
is it still working?
I am on app.wallabag.it and I can't retrieve the full articles.

**Comment by HolgerAusB** [comment]
Hi @Gullumluvl. Unfortunately @nicosomb didn't find the time yet, to install the release on app.wallabag.it which contains the changes. The version there is stuck on 2.6.6 from Oct. 3 2023. Don't know, how TontonMollo could have success with it.

As I am not a subscriber of monde-d, I can't check that.

**Comment by Gullumluvl** [comment]
Alright, then it's just a matter of waiting a little, no problem.

Just a perhaps naive question, but in cases like this, would it be feasible to implement a fallback solution, such as loading the article from a saved html file, or directly from the browser tab?

**Comment by yomli** [comment]
@Gullumluvl In theory, Wallabag can retrieve any html file if you feed it a public URL. So, saving the article as html, then uploading said html to a webhost may work. (I may or may not have done just that to get articles from a RSS feed…) Some processing may be needed, though, since Wallabag use a generic content extractor before switching to a URL-based specific one if I recall correctly.

TL;DR: try to upload your saved html file somewhere and use that URL.

**Comment by HolgerAusB** [comment]
@Gullumluvl you can use '[wallabagger](https://github.com/wallabag/wallabagger)', a plugin for desktop browsers. This catches the html that is already rendered by your browser and send it completely to wallabag. That might help for the missing user_agent in the old config. But wallabag will also use this (old) config to grab the good parts of the html.

But you should definitely give wallabagger a try, that plugin is very helpful for many websites, which cannot be grabed otherwise.

After installing you have to activate the option 'Content fetching from the browser' in the settings of this plugin.
